doc/design/move_and_copy_helper.md

```diff
@@ -0,0 +1,131 @@
+# Move And Copy Helper
+
+## Summary and problem description
+
+Currently, some classes in our codebase use assignment in the constructor (ctor) to avoid code duplication, as seen in issue [#1517](https://github.com/eclipse-iceoryx/iceoryx/issues/1517) . However, it's noted in issue [#1686](https://github.com/eclipse-iceoryx/iceoryx/issues/1686) that self-assignment cannot happen in the ctor. Therefore, an alternative approach should be adopted to replace the current implementation.
+
+One method involves placing the shared logical code within a separate function, hereafter referred to as `copy_and_move_impl`. This function is then called within the ctor and assignment operations, using template parameters to specify the method of member initialization.
+
+## Premise
+
+When a class performs highly consistent tasks during construction (ctor) and assignment, implementers often extract the repetitive parts into a separate `copy_and_move_impl`. However, there are inevitably minor differences between move, copy, and ctor, assignment (for example, during ctor, specific variables need to be initialized to avoid garbage values affecting subsequent behaviors, etc...). In these cases, the `MoveAndCopyHelper` can be used to simplify the code.
+
+## Terminology
+
+- `MoveAndCopyOperations`: an enum class, including
+
+    - `CopyConstructor`
+    - `MoveConstructor`
+    - `CopyAssignment`
+    - `MoveAssignment`
+
+## Goals
+
+The purpose of the `MoveAndCopyHelper` is:
+- To allow special constructors to indicate the source of the call (e.g., CopyConstructor) when calling `copy_and_move_impl`.
+- To provide a consistent way of expressing data exchange in `copy_and_move_impl`.
+- To offer a way to ignore `MoveAndCopyOperations` and forcibly use ctor or assignment, facilitating users' needs in special circumstances.
+
+## Design
+
+- Provide the enum class `MoveAndCopyOperations` to describe the expected behavior of member variables during initialization (e.g., using copy or move). Passing this enum as a template parameter to `copy_and_move_impl` ensures whether special constructors are called as expected.
+- Provide functions like `transfer` to ensure that the exchange process of member variables conforms to `MoveAndCopyOperations`.
+
+### Code example
+
+```cpp
+// header
+
+#include "iox/move_and_copy_helper.hpp"
+
+#define TMP 1024
+
+class Test
+{
+  public:
+    Test();
+    Test(const Test&);
+    Test(Test&&);
+    Test& operator=(const Test&);
+    Test& operator=(Test&&);
+
+    uint64_t size();
+
+  private:
+    int m_data[TMP];
+    uint64_t m_size{0};
+}
+
+// implementation
+Test::Test(const Test& rhs)
+{
+    copy_and_move_impl<MoveAndCopyOperations::CopyConstructor>(rhs);
+}
+
+Test::Test(Test&& rhs)
+{
+     copy_and_move_impl<MoveAndCopyOperations::MoveConstructor>(rhs);
+}
+
+Test& Test::operator=(const Test& rhs)
+{
+    if (this != &rhs)
+    {
+        copy_and_move_impl<MoveAndCopyOperations::CopyAssignment>(rhs);
+    }
```

Comment:
I believe that performing an early check helps to avoid unnecessary entry into the `copy_and_move_impl`, which would waste a bit of stack switching. If one wants to implement self-assignment check within `copy_and_move_impl`, it can be done using:

```cpp
if constexpr (!is_ctor)
{
    if (this == &rhs)
    {
        return;
    }      
}
```

Additionally, if `copy_and_move_impl` performs more runtime checks before self-assignment check, there would be a slight performance degradation."